state-synchronizer/src/peer_manager.rs

```diff
@@ -178,24 +210,43 @@ impl PeerManager {
     }
 
     pub fn process_request(&mut self, version: u64, peer_id: PeerId) {
-        self.requests.insert(version, (peer_id, SystemTime::now()));
+        if let Some(prev_request) = self.requests.get_mut(&version) {
+            prev_request.last_request_peer = peer_id;
+            prev_request.last_request_time = SystemTime::now();
+        } else {
+            self.requests
+                .insert(version, ChunkRequestInfo::new(version, peer_id));
+        }
+        if let Some(peer) = self.peers.get_mut(&peer_id) {
+            peer.pending_requests.push_back(version);
+            if peer.pending_requests.len() > MAX_PENDING_REQUESTS_PER_PEER {
+                peer.pending_requests.pop_front();
+            }
+        }
+    }
+
+    pub fn get_last_request_time(&self, version: u64) -> Option<SystemTime> {
+        self.requests
+            .get(&version)
+            .map(|req_info| req_info.last_request_time)
     }
 
-    pub fn get_request_time(&self, version: u64) -> Option<SystemTime> {
-        self.requests.get(&version).map(|(_, tst)| tst).cloned()
+    pub fn get_first_request_time(&self, version: u64) -> Option<SystemTime> {
+        self.requests
+            .get(&version)
+            .map(|req_info| req_info.first_request_time)
     }
 
     pub fn process_response(&mut self, version: u64, peer_id: PeerId) {
-        if let Some((id, _)) = self.requests.get(&version) {
-            if *id == peer_id {
-                self.requests.remove(&version);
-            }
+        if let Some(peer) = self.peers.get_mut(&peer_id) {
+            // cleanup pending requests prefixing the given version
+            peer.pending_requests.retain(|&v| v > version);
         }
     }
 
     pub fn has_requested(&self, version: u64, peer_id: PeerId) -> bool {
-        if let Some((id, _)) = self.requests.get(&version) {
-            return *id == peer_id;
+        if let Some(peer) = self.peers.get(&peer_id) {
+            return peer.pending_requests.contains(&version);
```

Comment:
does this mean if the request comes back out of order we'll penalize the peer?

if we request 10-20 from p1, and timeout out then request from p2 successfully then 20-30 from p1, then we'll clear the pending_requests. if 10-20 from p1 then comes back, has_requested will fail

Reply:
`pending_requests` of a peer are cleared only upon the response from that very peer. I might have misunderstood your example but in general this solution would penalize p1 only if it would first return a response for chunk 2 and only then return a response for chunk 1. I think it's ok to do so, otherwise it might be hard to GC pending request metadata.